twake/frontend/src/app/scenes/Client/ChannelsBar/Modals/SearchListManager.ts

```diff
@@ -0,0 +1,218 @@
+
+export type GenericChannel = {
+  type: 'user' | 'direct' | 'workspace';
+  sortString: string;
+  filterString: string;
+  lastActivity?: number;
+  resource: UserType | ChannelResource;
+};
+
+class SearchListManager {
+  private workspaceChannels: GenericChannel[];
+  private directChannels: GenericChannel[];
+  private users: GenericChannel[];
+  public list: GenericChannel[];
+
+  constructor() {
+    this.workspaceChannels = [];
+    this.directChannels = [];
+    this.users = [];
+    this.list = [];
+  }
+
+  public bind(search: string): void {
+    const { workspaceId, companyId } = RouterServices.getStateFromRoute();
+    // Path
+    const workspaceChannelsPath = `/channels/v1/companies/${companyId}/workspaces/${workspaceId}/channels/`;
+    const directChannelsPath = `/channels/v1/companies/${companyId}/workspaces/direct/channels/::mine`;
+    const mineWorkspaceChannelsPath = `/channels/v1/companies/${companyId}/workspaces/${workspaceId}/channels/::mine`;
+
+    // Resources
+    const workspaceChannelsResources = this.find({
+      collectionPath: workspaceChannelsPath,
+    }) as ChannelResource[];
+
+    const directChannelsResources = this.find({
+      collectionPath: directChannelsPath,
+    }) as ChannelResource[];
+
+    const mineWorkspaceChannelsResources = this.find({
+      collectionPath: mineWorkspaceChannelsPath,
+    }) as ChannelResource[];
+
+    const usersSearched = this.find({ search }) as UserType[];
+
+    // Filters
+    this.workspaceChannels = this.filterWorkspaceChannels({
+      channels: workspaceChannelsResources,
+      mineWorkspaceChannels: mineWorkspaceChannelsResources,
+    });
+    this.directChannels = this.filterDirectChannels({ channels: directChannelsResources });
+    this.users = this.filterUsers({ users: usersSearched });
+
+    // Concat list
+    this.list = [...this.workspaceChannels, ...this.directChannels, ...this.users];
+
+    this.removeDuplicate();
+
+    this.list = this.list
+      .filter(({ filterString }) => filterString.toUpperCase().indexOf(search.toUpperCase()) > -1)
+      .sort((a, b) => (b.lastActivity || 0) - (a.lastActivity || 0));
+  }
+
+  private find({
+    collectionPath,
+    search,
+  }: {
+    collectionPath?: string;
+    search?: string;
+  }): (UserType | ChannelResource)[] {
+    if (collectionPath?.length) {
+      const collection = Collection.get(collectionPath, ChannelResource);
+      return collection.find({});
+    } else {
+      let users: UserType[] = [];
+
+      this.searchUsers(res => (users = res), search || '');
+
+      return users;
+    }
+  }
+
+  private async searchUsers(cb: (res: UserType[]) => UserType[], text: string) {
+    if (text) {
+      return UsersService.search(
+        Strings.removeAccents(text),
+        {
+          group_id: Workspaces.currentGroupId,
+        },
+        (res: UserType[]) => cb(res.filter((el: UserType) => !!el)),
+      );
+    }
+  }
```

Comment:
This is a bad pattern to give `users` to the `searchUsers` function here. What about returning a Promise in searchUsers?
Also your `find` function is sync and you have callback here. Is it expected?